transform/transformers/common_software/pck_transformer.py

```diff
@@ -283,6 +387,57 @@ def calculate_total_playback(self):
+
+    def _compute_single_total_qsi_totals(self):
+        """
+        Calculates the start and end stock values for QSI (Stocks).  Saves these to qcode 65 and 66 respectively except for a few types
+        that have the qcode for the totals defined in the mapping.
+        """
+        instrument_id = self.response['collection']['instrument_id']
+        try:
+            start_questions = self.qsi_questions[instrument_id]['start']
+            end_questions = self.qsi_questions[instrument_id]['end']
+            start_total_qcode = self.qsi_questions[instrument_id].get('start_total_qcode', '65')
+            end_total_qcode = self.qsi_questions[instrument_id].get('end_total_qcode', '66')
+        except KeyError:
+            logger.exception("Missing key from mapping.  Is the mapping for the formtype correct?", formtype=instrument_id)
+            raise
+
+        start_total = sum(Decimal(value) for q_code, value in self.data.items() if q_code in start_questions)
+        end_total = sum(Decimal(value) for q_code, value in self.data.items() if q_code in end_questions)
+        self.data[start_total_qcode] = str(start_total)
+        self.data[end_total_qcode] = str(end_total)
+
+    def _compute_multiple_total_qsi_totals(self):
+        """
+        Calculates the start and end stock values for QSI (Stocks).  This is used for the two formtypes that have two
+        totals to calculate. Saves the total values to 298 and 299 for the non-dwelling questions and 398 and 399 for the
+        dwelling questions.
+        """
+        instrument_id = self.response['collection']['instrument_id']
+        try:
+            non_dwelling_start_questions = self.qsi_questions[instrument_id]['non_dwelling_questions_start']
+            non_dwelling_end_questions = self.qsi_questions[instrument_id]['non_dwelling_questions_end']
+            dwelling_start_questions = self.qsi_questions[instrument_id]['dwelling_questions_start']
+            dwelling_end_questions = self.qsi_questions[instrument_id]['dwelling_questions_end']
+        except KeyError:
+            logger.exception("Missing key from mapping.  Is the mapping for the formtype correct?", formtype=instrument_id)
+            raise
+
+        non_dwelling_start_total = sum(Decimal(value) for q_code, value in self.data.items() if q_code in non_dwelling_start_questions)
+        non_dwelling_end_total = sum(Decimal(value) for q_code, value in self.data.items() if q_code in non_dwelling_end_questions)
+        dwelling_start_total = sum(Decimal(value) for q_code, value in self.data.items() if q_code in dwelling_start_questions)
+        dwelling_end_total = sum(Decimal(value) for q_code, value in self.data.items() if q_code in dwelling_end_questions)
+        self.data['298'] = str(non_dwelling_start_total)
+        self.data['299'] = str(non_dwelling_end_total)
+        self.data['398'] = str(dwelling_start_total)
+        self.data['399'] = str(dwelling_end_total)
```

Comment:
I'm not so familiar with this software, but this practice of introducing large impure functions on a per-survey basis would generally be considered overly complex and thus prone to unexpected actions.

Out of scope for this ticket, but is there potential for creating smaller pure functions for selective use within the custom functions for each survey, to increase code reuse, and increase testability of that reuse? Also seems like an ETL transformation data type might be useful to genericise this further.

You're best placed to know if this is doable at some point, but if it is, it could make some of the repetition involved in these onboardings a lot easier in future - however I concede that it's a big step, and the benefit of it would have to be carefully considered.  There's an element of "If it ain't broke" involved.